File: Daily/sbgo/Leetcode1178-猜灯谜.go
package sbgo

func findNumOfValidWords(words []string, puzzles []string) []int {
	freq := make(map[int]int)
	for _, word := range words {
		mask := 0
		for _, char := range word {
			mask |= (1 << (char - 'a'))
		}
		freq[mask]++
	}

	ans := []int{}

	for _, puzzle := range puzzles {
		mask := 0
		for _, char := range puzzle {
			mask |= (1 << (char - 'a'))
		}
		total := 0
		subset := mask
		for subset != 0 {
			if 1<<(puzzle[0]-'a')&subset != 0 {
				total += freq[subset]
			}
			subset = (subset - 1) & mask
		}
		ans = append(ans, total)
	}
	return ans
}

File: Leetcode/PY/Leetcode1178-猜灯谜.py
from typing import List
from collections import defaultdict
class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        freq = defaultdict(int)
        for word in words:
            mask = 0
            for char in word:
                mask |= (1 << (ord(char) - ord('a')))
            freq[mask] += 1
        
        ans = list()
        
        for puzzle in puzzles:
            mask = 0
            total = 0
            for char in puzzle:
                mask |= (1 << (ord(char) - ord("a")))
            subset = mask
            while subset != 0:
                if 1 << (ord(puzzle[0]) - ord('a')) & subset != 0:
                    if subset in freq:
                        total += freq[subset]
                subset = (subset - 1) & mask
            ans.append(total)
            
        return ans